domain/application/state/resource.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+
+	coreresource "github.com/juju/juju/core/resource"
+	"github.com/juju/juju/domain/application"
+	"github.com/juju/juju/internal/errors"
+)
+
+// buildResourceToAdd creates resources to add based on provided app and charm
+// resources.
+// Returns a slice of resourceToAdd and an error if any issues occur during
+// creation. Validates app resources against charm resources, warning if
+// mismatches are found.
+func (st *State) buildResourceToAdd(appUUID, charmUUID string,
+	appResources []application.AddApplicationResourceArg) (
+	[]resourceToAdd, error) {
+	resources := make([]resourceToAdd, 0, len(appResources))
+	now := st.clock.Now()
+	for _, r := range appResources {
+		uuid, err := coreresource.NewUUID()
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		resources = append(resources, resourceToAdd{
+			UUID:            uuid.String(),
+			ApplicationUUID: appUUID,
+			CharmUUID:       charmUUID,
+			Name:            r.Name,
+			Revision:        r.Revision,
+			Origin:          r.Origin.String(),
+			State:           coreresource.StateAvailable.String(),
+			CreatedAt:       now,
+		})
+	}
+	return resources, nil
+}
+
+// insertResources constructs a transaction to insert resources into the
+// database. It returns a function which, when executed, inserts resources and
+// links them to applications.
+func (st *State) insertResources(ctx context.Context, tx *sqlair.TX, appDetails applicationDetails, appResources []application.AddApplicationResourceArg) error {
+
+	resources, err := st.buildResourceToAdd(appDetails.UUID.String(), appDetails.CharmID, appResources)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Prepare SQL statement to get the origin of a specific resource
+	getOriginIDStmt, err := st.Prepare(`
+SELECT id AS &resourceToAdd.origin_type_id
+FROM   resource_origin_type
+WHERE  name = $resourceToAdd.origin_type_name`, resourceToAdd{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Prepare SQL statement to get the state of a specific resource.
+	getStateIDStmt, err := st.Prepare(`
+SELECT id AS &resourceToAdd.state_id
+FROM   resource_state
+WHERE  name = $resourceToAdd.state_name`, resourceToAdd{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Prepare SQL statement to insert the resource.
+	insertStmt, err := st.Prepare(`
+INSERT INTO resource (uuid, charm_uuid, charm_resource_name, revision, 
+origin_type_id, state_id, created_at)
+VALUES ($resourceToAdd.*)`, resourceToAdd{})
+	if err != nil {
+		return errors.Capture(err)
+	}
```

Comment:
This can be done in one step:

```sql
INSERT INTO resource (uuid, charm_uuid, charm_resource_name, revision, 
       origin_type_id, state_id, created_at)
SELECT $resourceToAdd.uuid,
       $resourceToAdd.charm_uuid,
       $resourceToAdd.charm_resource_name,
       $resourceToAdd.revision,
       rot.id,
       rs.id,
       $resourceToAdd.created_at
FROM   resource_origin_type rot,
       resource_state rs
WHERE  rot.name = $resourceToAdd.origin_type_name
AND    rs.name = $resourceToAdd.state_name
```

Which allows for redefining the structs so they aren't overloaded:
```
type resourceInsert struct {
 	UUID string `db:"uuid"`
 	CharmUUID       string `db:"charm_uuid"`
 	Name            string `db:"charm_resource_name"`
 	Revision     *int      `db:"revision"`
 	Origin       string    `db:"origin_type_name"`
 	State        string    `db:"state_name"`
 	CreatedAt    time.Time `db:"created_at"`
 }

type linkResourceApplication struct {
	UUID string `db:"uuid"`
 	ApplicationUUID string `db:"application_uuid"`
}

```

Reply:
Oh, nice ! Thanks!

Reply:
Joe showed me this for permissions. :-)